File: sample_files_all_types/sample.kt
package main.prd

/**
 * PRD Management System - Kotlin Implementation
 * Version: 1.2.0 | Last Updated: July 25, 2025
 */

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.collections.*

@Serializable
enum class PRDStatus(val value: Int, val displayName: String) {
    DRAFT(0, "Draft"),
    IN_REVIEW(1, "In Review"),
    APPROVED(2, "Approved"),
    IN_DEVELOPMENT(3, "In Development"),
    TESTING(4, "Testing"),
    IMPLEMENTED(5, "Implemented"),
    ARCHIVED(6, "Archived");

    companion object {
        fun fromValue(value: Int): PRDStatus = values().find { it.value == value } ?: DRAFT
    }
}

@Serializable
enum class Priority(val value: Int, val displayName: String, val colorCode: String) {
    LOW(1, "Low", "#28a745"),
    MEDIUM(2, "Medium", "#ffc107"),
    HIGH(3, "High", "#fd7e14"),
    CRITICAL(4, "Critical", "#dc3545");

    companion object {
        fun fromValue(value: Int): Priority = values().find { it.value == value } ?: MEDIUM
    }
}

@Serializable
data class PRD(
    val id: String = generateId(),
    var title: String,
    var description: String,
    var author: String,
    var status: PRDStatus = PRDStatus.DRAFT,
    var priority: Priority = Priority.MEDIUM,
    @Serializable(with = LocalDateTimeSerializer::class)
    val createdAt: LocalDateTime = LocalDateTime.now(),
    @Serializable(with = LocalDateTimeSerializer::class)
    var updatedAt: LocalDateTime = LocalDateTime.now(),
    var completionPercentage: Int = 0,
    var tags: MutableList<String> = mutableListOf()
) {
    companion object {
        private fun generateId(): String {
            val timestamp = System.currentTimeMillis()
            val random = (1000..9999).random()
            return "PRD-$timestamp-$random"
        }
    }

    fun updateStatus(newStatus: PRDStatus) {
        status = newStatus
        updatedAt = LocalDateTime.now()
    }

    fun setCompletionPercentage(percentage: Int) {
        completionPercentage = percentage.coerceIn(0, 100)
        updatedAt = LocalDateTime.now()
    }

    fun addTag(tag: String) {
        val cleanTag = tag.trim().lowercase()
        if (cleanTag.isNotBlank() && !tags.contains(cleanTag)) {
            tags.add(cleanTag)
            updatedAt = LocalDateTime.now()
        }
    }

    fun setPriority(newPriority: Priority) {
        priority = newPriority
        updatedAt = LocalDateTime.now()
    }

    val progressDescription: String
        get() = "$completionPercentage% complete"

    val statusIcon: String
        get() = when (status) {
            PRDStatus.DRAFT -> "📝"
            PRDStatus.IN_REVIEW -> "👁️"
            PRDStatus.APPROVED -> "✅"
            PRDStatus.IN_DEVELOPMENT -> "🔨"
            PRDStatus.TESTING -> "🧪"
            PRDStatus.IMPLEMENTED -> "⭐"
            PRDStatus.ARCHIVED -> "📦"
        }

    override fun toString(): String {
        return "PRD{ID='$id', Title='$title', Status=${status.displayName}, Completion=$completionPercentage%}"
    }
}

@Serializable
data class Analytics(
    val totalPRDs: Int,
    val statusCounts: Map<String, Int>,
    val priorityCounts: Map<String, Int>,
    val averageCompletion: Double,
    val topAuthors: Map<String, Int>,
    val tagFrequency: Map<String, Int>,
    @Serializable(with = LocalDateTimeSerializer::class)
    val lastUpdated: LocalDateTime
) {
    val mostUsedTags: List<Pair<String, Int>>
        get() = tagFrequency.toList().sortedByDescending { it.second }

    val topContributors: List<Pair<String, Int>>
        get() = topAuthors.toList().sortedByDescending { it.second }
}

object LocalDateTimeSerializer : KSerializer<LocalDateTime> {
    override val descriptor = PrimitiveSerialDescriptor("LocalDateTime", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: LocalDateTime) {
        encoder.encodeString(value.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
    }

    override fun deserialize(decoder: Decoder): LocalDateTime {
        return LocalDateTime.parse(decoder.decodeString(), DateTimeFormatter.ISO_LOCAL_DATE_TIME)
    }
}

class PRDManager {
    private val prds = mutableListOf<PRD>()
    private val prdIndex = mutableMapOf<String, Int>()

    // Core Operations
    fun createPRD(title: String, description: String, author: String): String {
        val prd = PRD(title = title, description = description, author = author)
        prdIndex[prd.id] = prds.size
        prds.add(prd)
        
        println("PRD created successfully: ${prd.id}")
        return prd.id
    }

    fun getPRD(id: String): PRD? {
        val index = prdIndex[id] ?: return null
        return prds.getOrNull(index)
    }

    fun getAllPRDs(): List<PRD> = prds.toList()

    fun getPRDsByStatus(status: PRDStatus): List<PRD> = prds.filter { it.status == status }

    fun getPRDsByPriority(priority: Priority): List<PRD> = prds.filter { it.priority == priority }

    fun searchPRDs(searchTerm: String): List<PRD> {
        val lowercaseTerm = searchTerm.lowercase()
        return prds.filter { prd ->
            prd.title.lowercase().contains(lowercaseTerm) ||
            prd.description.lowercase().contains(lowercaseTerm) ||
            prd.tags.any { it.contains(lowercaseTerm) }
        }
    }

    fun updatePRDStatus(id: String, newStatus: PRDStatus): Boolean {
        val prd = getPRD(id) ?: return false
        prd.updateStatus(newStatus)
        println("PRD $id status updated to: ${newStatus.displayName}")
        return true
    }

    fun updatePRDCompletion(id: String, percentage: Int): Boolean {
        val prd = getPRD(id) ?: return false
        prd.setCompletionPercentage(percentage)
        return true
    }

    // Analytics
    fun generateAnalytics(): Analytics {
        val totalPRDs = prds.size
        val statusCounts = prds.groupingBy { it.status.displayName }.eachCount()
        val priorityCounts = prds.groupingBy { it.priority.displayName }.eachCount()
        val authorCounts = prds.groupingBy { it.author }.eachCount()
        
        val tagCounts = mutableMapOf<String, Int>()
        prds.forEach { prd ->
            prd.tags.forEach { tag ->
                tagCounts[tag] = tagCounts.getOrDefault(tag, 0) + 1
            }
        }

        val averageCompletion = if (totalPRDs > 0) {
            prds.sumOf { it.completionPercentage }.toDouble() / totalPRDs
        } else 0.0

        return Analytics(
            totalPRDs = totalPRDs,
            statusCounts = statusCounts,
            priorityCounts = priorityCounts,
            averageCompletion = averageCompletion,
            topAuthors = authorCounts,
            tagFrequency = tagCounts,
            lastUpdated = LocalDateTime.now()
        )
    }

    // Data Management
    fun exportToJSON(): String {
        val json = Json { 
            prettyPrint = true
            encodeDefaults = true
        }
        return json.encodeToString(prds)
    }

    fun importFromJSON(jsonString: String): Boolean {
        return try {
            val json = Json { ignoreUnknownKeys = true }
            val importedPRDs = json.decodeFromString<List<PRD>>(jsonString)
            
            prds.clear()
            prdIndex.clear()
            
            importedPRDs.forEach { prd ->
                prdIndex[prd.id] = prds.size
                prds.add(prd)
            }
            true
        } catch (e: Exception) {
            println("Error importing JSON: ${e.message}")
            false
        }
    }

    // Dashboard
    fun printDashboard() {
        println("\n${"=".repeat(60)}")
        println("PRD MANAGEMENT SYSTEM - DASHBOARD")
        println("${"=".repeat(60)}")

        val analytics = generateAnalytics()

        println("Total PRDs: ${analytics.totalPRDs}")
        println("Average Completion: ${"%.1f".format(analytics.averageCompletion)}%")

        println("\nStatus Distribution:")
        analytics.statusCounts.toSortedMap().forEach { (status, count) ->
            println("  $status: $count")
        }

        println("\nPriority Distribution:")
        analytics.priorityCounts.toSortedMap().forEach { (priority, count) ->
            println("  $priority: $count")
        }

        println("\nTop Authors:")
        analytics.topContributors.take(5).forEach { (author, count) ->
            println("  $author: $count PRDs")
        }

        println("\nMost Used Tags:")
        analytics.mostUsedTags.take(5).forEach { (tag, count) ->
            println("  #$tag: $count times")
        }

        println("\nRecent PRDs:")
        prds.sortedByDescending { it.updatedAt }.take(5).forEach { prd ->
            println("  $prd")
        }
    }

    // Sample Data
    fun loadSampleData() {
        val sampleData = listOf(
            Triple("User Authentication System", "Implement secure login and registration", "Dev Team") to listOf("security", "authentication"),
            Triple("Dark Mode Theme", "Add dark theme option for better UX", "UX Team") to listOf("ui", "theme"),
            Triple("Payment Gateway Integration", "Integrate secure payment processing", "Product Team") to listOf("payment", "integration"),
            Triple("API Rate Limiting", "Implement API rate limiting for security", "Backend Team") to listOf("api", "security"),
            Triple("Mobile App Redesign", "Complete redesign of mobile application", "Design Team") to listOf("mobile", "design"),
            Triple("Real-time Notifications", "Add real-time notification system", "Full Stack Team") to listOf("notifications", "realtime"),
            Triple("Performance Optimization", "Optimize database queries and caching", "Database Team") to listOf("performance", "database"),
            Triple("Multi-language Support", "Add internationalization support", "Localization Team") to listOf("i18n", "localization"),
            Triple("Biometric Authentication", "Add fingerprint and face recognition", "Security Team") to listOf("biometric", "android"),
            Triple("Offline Mode", "Enable app functionality without internet", "Mobile Team") to listOf("offline", "sync")
        )

        sampleData.forEach { (triple, tags) ->
            val (title, description, author) = triple
            val id = createPRD(title, description, author)
            val prd = getPRD(id)
            tags.forEach { prd?.addTag(it) }
        }

        // Update some statuses and priorities for variety
        if (prds.size >= 10) {
            updatePRDStatus(prds[1].id, PRDStatus.IN_REVIEW)
            updatePRDStatus(prds[2].id, PRDStatus.APPROVED)
            updatePRDStatus(prds[3].id, PRDStatus.IN_DEVELOPMENT)
            updatePRDStatus(prds[4].id, PRDStatus.TESTING)
            updatePRDStatus(prds[5].id, PRDStatus.IMPLEMENTED)

            prds[2].setPriority(Priority.HIGH)
            prds[3].setPriority(Priority.CRITICAL)
            prds[8].setPriority(Priority.HIGH)

            updatePRDCompletion(prds[3].id, 65)
            updatePRDCompletion(prds[4].id, 90)
            updatePRDCompletion(prds[5].id, 100)
        }
    }

    // Utility Methods
    fun getCompletionStats(): Triple<Int, Int, Double> {
        if (prds.isEmpty()) return Triple(0, 0, 0.0)

        val completions = prds.map { it.completionPercentage }
        val min = completions.minOrNull() ?: 0
        val max = completions.maxOrNull() ?: 0
        val average = completions.average()

        return Triple(min, max, average)
    }

    fun getPRDsNeedingAttention(): List<PRD> {
        return prds.filter { prd ->
            (prd.status == PRDStatus.IN_DEVELOPMENT && prd.completionPercentage < 50) ||
            (prd.priority == Priority.CRITICAL && prd.status == PRDStatus.DRAFT) ||
            (prd.status == PRDStatus.TESTING && prd.completionPercentage < 80)
        }
    }

    fun getStatusProgressReport(): Map<PRDStatus, Double> {
        return PRDStatus.values().associateWith { status ->
            val prdsWithStatus = getPRDsByStatus(status)
            if (prdsWithStatus.isEmpty()) 0.0
            else prdsWithStatus.map { it.completionPercentage }.average()
        }
    }
}

// Demo Implementation
fun main() {
    println("PRD Management System v1.2.0 - Kotlin Implementation")
    println("====================================================")

    // Initialize the manager
    val manager = PRDManager()

    // Load sample data
    manager.loadSampleData()

    // Display dashboard
    manager.printDashboard()

    // Demo operations
    println("\n${"=".repeat(60)}")
    println("DEMO OPERATIONS")
    println("${"=".repeat(60)}")

    // Search demo
    val searchResults = manager.searchPRDs("authentication")
    println("\nSearching for 'authentication' related PRDs:")
    searchResults.forEach { prd ->
        println("  Found: $prd")
    }

    // Filter by status demo
    val draftPRDs = manager.getPRDsByStatus(PRDStatus.DRAFT)
    println("\nDraft PRDs (${draftPRDs.size}):")
    draftPRDs.forEach { prd ->
        println("  $prd")
    }

    // Priority filtering demo
    val criticalPRDs = manager.getPRDsByPriority(Priority.CRITICAL)
    println("\nCritical Priority PRDs (${criticalPRDs.size}):")
    criticalPRDs.forEach { prd ->
        println("  ${prd.statusIcon} $prd")
    }

    // PRDs needing attention
    val needingAttention = manager.getPRDsNeedingAttention()
    println("\nPRDs Needing Attention (${needingAttention.size}):")
    needingAttention.forEach { prd ->
        println("  ${prd.statusIcon} $prd - ${prd.priority.displayName} priority")
    }

    // Completion statistics
    val (min, max, average) = manager.getCompletionStats()
    println("\nCompletion Statistics:")
    println("  Minimum: $min%")
    println("  Maximum: $max%")
    println("  Average: ${"%.1f".format(average)}%")

    // Status progress report
    println("\nStatus Progress Report:")
    val statusReport = manager.getStatusProgressReport()
    statusReport.forEach { (status, avgCompletion) ->
        println("  ${status.displayName}: ${"%.1f".format(avgCompletion)}% average completion")
    }

    // Export demo
    println("\nExporting PRD data to JSON...")
    try {
        val jsonData = manager.exportToJSON()
        println("Export completed. JSON length: ${jsonData.length} characters")
        
        // Test import
        val testManager = PRDManager()
        if (testManager.importFromJSON(jsonData)) {
            println("JSON import test: Success!")
        }
    } catch (e: Exception) {
        println("Error exporting to JSON: ${e.message}")
    }

    println("\nKotlin PRD Management System demonstration completed!")
}

File: sample_files_all_types/sample.cs
/**
 * PRD Management System - C# Implementation
 * Version: 1.2.0 | Last Updated: July 25, 2025
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace PRDManagement
{
    public enum PRDStatus
    {
        Draft = 0,
        InReview = 1,
        Approved = 2,
        InDevelopment = 3,
        Testing = 4,
        Implemented = 5,
        Archived = 6
    }

    public enum Priority
    {
        Low = 1,
        Medium = 2,
        High = 3,
        Critical = 4
    }

    public class PRD
    {
        public string Id { get; private set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public string Author { get; set; }
        public PRDStatus Status { get; set; }
        public Priority Priority { get; set; }
        public DateTime CreatedAt { get; private set; }
        public DateTime UpdatedAt { get; set; }
        public int CompletionPercentage { get; set; }
        public List<string> Tags { get; set; }

        public PRD(string title, string description, string author)
        {
            Id = GenerateId();
            Title = title;
            Description = description;
            Author = author;
            Status = PRDStatus.Draft;
            Priority = Priority.Medium;
            CreatedAt = DateTime.UtcNow;
            UpdatedAt = DateTime.UtcNow;
            CompletionPercentage = 0;
            Tags = new List<string>();
        }

        private string GenerateId()
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var random = new Random().Next(1000, 9999);
            return $"PRD-{timestamp}-{random}";
        }

        public void UpdateStatus(PRDStatus newStatus)
        {
            Status = newStatus;
            UpdatedAt = DateTime.UtcNow;
        }

        public void SetCompletionPercentage(int percentage)
        {
            CompletionPercentage = Math.Max(0, Math.Min(100, percentage));
            UpdatedAt = DateTime.UtcNow;
        }

        public override string ToString()
        {
            return $"PRD{{Id='{Id}', Title='{Title}', Status={Status}, Completion={CompletionPercentage}%}}";
        }
    }

    public class PRDManager
    {
        private readonly List<PRD> _prds;
        private readonly Dictionary<string, PRD> _prdIndex;

        public PRDManager()
        {
            _prds = new List<PRD>();
            _prdIndex = new Dictionary<string, PRD>();
        }

        public string CreatePRD(string title, string description, string author)
        {
            var prd = new PRD(title, description, author);
            _prds.Add(prd);
            _prdIndex[prd.Id] = prd;

            Console.WriteLine($"PRD created successfully: {prd.Id}");
            return prd.Id;
        }

        public PRD GetPRD(string id)
        {
            return _prdIndex.TryGetValue(id, out var prd) ? prd : null;
        }

        public List<PRD> GetAllPRDs()
        {
            return _prds.ToList();
        }

        public List<PRD> GetPRDsByStatus(PRDStatus status)
        {
            return _prds.Where(prd => prd.Status == status).ToList();
        }

        public List<PRD> GetPRDsByPriority(Priority priority)
        {
            return _prds.Where(prd => prd.Priority == priority).ToList();
        }

        public List<PRD> SearchPRDs(string searchTerm)
        {
            var term = searchTerm.ToLowerInvariant();
            return _prds.Where(prd => 
                prd.Title.ToLowerInvariant().Contains(term) || 
                prd.Description.ToLowerInvariant().Contains(term) ||
                prd.Tags.Any(tag => tag.ToLowerInvariant().Contains(term))
            ).ToList();
        }

        public bool UpdatePRDStatus(string id, PRDStatus newStatus)
        {
            var prd = GetPRD(id);
            if (prd != null)
            {
                prd.UpdateStatus(newStatus);
                Console.WriteLine($"PRD {id} status updated to: {newStatus}");
                return true;
            }
            return false;
        }

        public Dictionary<string, object> GetStatistics()
        {
            var stats = new Dictionary<string, object>
            {
                ["TotalPRDs"] = _prds.Count,
                ["ByStatus"] = _prds.GroupBy(prd => prd.Status)
                                  .ToDictionary(g => g.Key.ToString(), g => g.Count()),
                ["ByPriority"] = _prds.GroupBy(prd => prd.Priority)
                                    .ToDictionary(g => g.Key.ToString(), g => g.Count()),
                ["AverageCompletion"] = _prds.Count > 0 ? _prds.Average(prd => prd.CompletionPercentage) : 0
            };

            return stats;
        }

        public string ExportToJson()
        {
            var options = new JsonSerializerOptions
            {
                WriteIndented = true,
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            };

            return JsonSerializer.Serialize(_prds, options);
        }

        public void PrintDashboard()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("PRD MANAGEMENT SYSTEM - DASHBOARD");
            Console.WriteLine(new string('=', 60));

            var stats = GetStatistics();
            Console.WriteLine($"Total PRDs: {stats["TotalPRDs"]}");
            Console.WriteLine($"Average Completion: {stats["AverageCompletion"]:F1}%");

            Console.WriteLine("\nStatus Distribution:");
            var statusCounts = (Dictionary<string, int>)stats["ByStatus"];
            foreach (var kvp in statusCounts)
            {
                Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
            }

            Console.WriteLine("\nPriority Distribution:");
            var priorityCounts = (Dictionary<string, int>)stats["ByPriority"];
            foreach (var kvp in priorityCounts)
            {
                Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
            }

            Console.WriteLine("\nRecent PRDs:");
            var recentPRDs = _prds.OrderByDescending(prd => prd.UpdatedAt).Take(5);
            foreach (var prd in recentPRDs)
            {
                Console.WriteLine($"  {prd}");
            }
        }

        public void LoadSampleData()
        {
            var sampleData = new[]
            {
                ("User Authentication System", "Implement secure login and registration", "Dev Team"),
                ("Dark Mode Theme", "Add dark theme option for better UX", "UX Team"),
                ("Payment Gateway Integration", "Integrate secure payment processing", "Product Team"),
                ("API Rate Limiting", "Implement API rate limiting for security", "Backend Team"),
                ("Mobile App Redesign", "Complete redesign of mobile application", "Design Team"),
                ("Real-time Notifications", "Add real-time notification system", "Full Stack Team"),
                ("Performance Optimization", "Optimize database queries and caching", "Database Team"),
                ("Multi-language Support", "Add internationalization support", "Localization Team")
            };

            foreach (var (title, description, author) in sampleData)
            {
                var id = CreatePRD(title, description, author);
                var prd = GetPRD(id);
                
                // Add some sample tags
                prd.Tags.AddRange(title.ToLowerInvariant().Split(' ').Take(2));
            }

            // Update some statuses for variety
            var allPRDs = GetAllPRDs();
            if (allPRDs.Count >= 8)
            {
                UpdatePRDStatus(allPRDs[1].Id, PRDStatus.InReview);
                UpdatePRDStatus(allPRDs[2].Id, PRDStatus.Approved);
                UpdatePRDStatus(allPRDs[3].Id, PRDStatus.InDevelopment);
                UpdatePRDStatus(allPRDs[4].Id, PRDStatus.Testing);
                UpdatePRDStatus(allPRDs[5].Id, PRDStatus.Implemented);
                
                allPRDs[3].SetCompletionPercentage(65);
                allPRDs[4].SetCompletionPercentage(90);
                allPRDs[5].SetCompletionPercentage(100);
            }
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("PRD Management System v1.2.0 - C# Implementation");
            Console.WriteLine("==================================================");

            var manager = new PRDManager();
            
            // Load sample data
            manager.LoadSampleData();

            // Display dashboard
            manager.PrintDashboard();

            // Demo some operations
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("DEMO OPERATIONS");
            Console.WriteLine(new string('=', 60));

            // Search demo
            var searchResults = manager.SearchPRDs("authentication");
            Console.WriteLine($"\nSearching for 'authentication' related PRDs:");
            searchResults.ForEach(prd => Console.WriteLine($"  Found: {prd}"));

            // Filter by status demo
            var draftPRDs = manager.GetPRDsByStatus(PRDStatus.Draft);
            Console.WriteLine($"\nDraft PRDs ({draftPRDs.Count}):");
            draftPRDs.ForEach(prd => Console.WriteLine($"  {prd}"));

            // Export demo
            Console.WriteLine("\nExporting PRD data to JSON...");
            var jsonData = manager.ExportToJson();
            Console.WriteLine($"Export completed. JSON length: {jsonData.Length} characters");

            Console.WriteLine("\nC# PRD Management System demonstration completed!");
        }
    }
}
